python_modules/dagster/dagster/_model/decorator.py

```diff
@@ -0,0 +1,246 @@
+from abc import ABC
+from functools import cached_property, partial
+from typing import (
+    Any,
+    Callable,
+    Mapping,
+    NamedTuple,
+    Optional,
+    Type,
+    TypeVar,
+    Union,
+    overload,
+)
+
+from typing_extensions import dataclass_transform
+
+from dagster._check import EvalContext, build_check_call
+from dagster._utils.cached_method import NT_CACHED_METHOD_CACHE_FIELD
+
+TType = TypeVar("TType", bound=Type)
+TVal = TypeVar("TVal")
+
+
+MODEL_MARKER = object()
+
+
+def _namedtuple_model_transform(
+    cls: TType,
+    *,
+    checked: bool,
+    enable_cached_method: bool,
+    with_new: bool,
+) -> TType:
+    """Transforms the input class in to one that inherits a generated NamedTuple base class
+    and:
+        * bans tuple methods that don't make sense for a model object
+        * creates a run time checked __new__  (optional)
+        * adds @cached_method support (optional).
+    """
+    field_set = {
+        **cls.__annotations__,
+        **({NT_CACHED_METHOD_CACHE_FIELD: Any} if enable_cached_method else {}),
+    }
+    base = NamedTuple(f"_{cls.__name__}", field_set.items())
+
+    if checked:
+        orig_new = base.__new__
+        checks_builder = LazyCheckBuilder(
+            cls.__annotations__,
+            EvalContext.capture_from_frame(2),  # this method, original decorator
+            1 if with_new else 0,
+        )
+
+        def __checked_new__(cls: TType, **kwargs):
+            for key, fn in checks_builder.checks.items():
+                fn(kwargs[key])
+
+            cache_fields = {NT_CACHED_METHOD_CACHE_FIELD: {}} if enable_cached_method else {}
+            return orig_new(cls, **kwargs, **cache_fields)
+
+        base.__new__ = __checked_new__  # type: ignore # unhappy with dropping positional args
+
+    if enable_cached_method:
+        base.__eq__ = __cached_method_safe_eq__
+        base.__hash__ = __cached_method_safe_hash__
+
+    return type(  # type: ignore
+        cls.__name__,
+        (cls, base),
+        {
+            "__iter__": _banned_iter,
+            "__getitem__": _banned_idx,
+            "__hidden_iter__": base.__iter__,
+            "__marker__": MODEL_MARKER,
```

Comment:
I think `__marker__` should be a less generic name. 

I do want to release this as `checkrepublic` one day so could do that :-)